### PR TITLE
Supporting rdbar/wrtbar changes for field watch

### PR DESCRIPTION
**Pull request by cathyzhyi**
This change adds field report helpers and jit resolve fields helpers, and add enableFieldWatch option

issue: #2812

Signed-off-by: Yi Zhang <yizhang@ca.ibm.com>

### CODE REVIEW

File: compiler/optimizer/PartialRedundancy.cpp [hunk 1]
```diff
@@ -230,6 +230,10 @@ int32_t TR_PartialRedundancy::perform()
    if (comp()->getProfilingMode() == JitProfiling && comp()->getHCRMode() != TR::none && _numProfilingsAllowed == 0)
       return 0;
 
+   // PRE is not supported for field watch yet
```
Comment (andrewcraik) [inline compiler/optimizer/PartialRedundancy.cpp]: I think this needs at least a reason - why isn't it supported, what is the TODO that needs to be addressed or the issue encountered?

**Comment by vijaysun-omr** [comment]
Can an issue be created to refactor the code such that the query is asked of the language "runtime" in some sense ?

**Comment by andrewcraik** [comment]
Ok so functionally I'm fine with this. The JIT helper names are a bit esoteric for my taste, but I can live with them. I do want a reason for PRE being disabled to be documented. I agree that we need an issue to factor the canGCandReturn query to facilitate language overrides since it may change between languages what the specific semantics of some nodes (like read and write barriers) are.

**Comment by cathyzhyi** [comment]
@andrewcraik @vijaysun-omr Thanks for the suggestion. I am keeping the OMR layer of canGCAndReturn as before as the default implementation and move the checking for `readbar` and `wrtbar` into the sub projects.

File: compiler/optimizer/PartialRedundancy.cpp (hunk 1 above)
Comment (cathyzhyi) [reply compiler/optimizer/PartialRedundancy.cpp]: added reason and TODO.

**Comment by cathyzhyi** [comment]
opened issue #3026 to further cleanup canGCAndReturn.

**Comment by vijaysun-omr** [comment]
What is so unique about PRE that it is disabled for readbar/writebar for field watch ? 
I mean, there are some other opts such as field privatizer for example that do some similar operations and I was wondering if those need to disabled too (or you have made the changes in those case) ?

**Comment by andrewcraik** [comment]
+1 to @vijaysun-omr - the comment isn't saying why it is disabled - the answer to the above question is what is needed in the comment in addition to addressing the other opts which do similar things.

**Comment by andrewcraik** [comment]
So it looks like this PR has now changed to just adding helpers and options - opt disables and other related changes have been removed. I'm fine with the helpers and options so review can continue anew on this. @cathyzhyi given the disable is removed can you just comment on if that is going to be in a separate PR or if it is un-needed now?

File: compiler/ras/Debug.cpp [hunk 2]
```diff
@@ -3912,6 +3912,14 @@ TR_Debug::getRuntimeHelperName(int32_t index)
          case TR_stackOverflow:             return "jitStackOverflow";
          case TR_reportMethodEnter:         return "jitReportMethodEnter";
          case TR_reportStaticMethodEnter:   return "jitReportStaticMethodEnter";
+         case TR_jitReportInstanceFieldRead: return "jitReportInstanceField";
+         case TR_jitReportStaticFieldRead:  return "jitReportStaticFieldRead";
+         case TR_jitReportInstanceFieldWrite:  return "jitReportInstanceWrite";
```
Comment (0xdaryl) [inline compiler/ras/Debug.cpp]: `"jitReportInstanceFieldWrite"` for consistency.

File: compiler/ras/Debug.cpp [hunk 3]
```diff
@@ -3912,6 +3912,14 @@ TR_Debug::getRuntimeHelperName(int32_t index)
          case TR_stackOverflow:             return "jitStackOverflow";
          case TR_reportMethodEnter:         return "jitReportMethodEnter";
          case TR_reportStaticMethodEnter:   return "jitReportStaticMethodEnter";
+         case TR_jitReportInstanceFieldRead: return "jitReportInstanceField";
```
Comment (0xdaryl) [inline compiler/ras/Debug.cpp]: `"jitReportInstanceFieldRead"`

File: compiler/ras/Debug.cpp (hunk 2 above)
Comment (cathyzhyi) [reply compiler/ras/Debug.cpp]: Thanks for pointing this out. Changed according to the suggestion.

File: compiler/ras/Debug.cpp (hunk 3 above)
Comment (cathyzhyi) [reply compiler/ras/Debug.cpp]: Thanks for pointing this out. Changed according to the suggestion.

**Comment by cathyzhyi** [comment]
> So it looks like this PR has now changed to just adding helpers and options - opt disables and other related changes have been removed. I'm fine with the helpers and options so review can continue anew on this. @cathyzhyi given the disable is removed can you just comment on if that is going to be in a separate PR or if it is un-needed now?

It will be addressed in a separate PR.

**Comment by andrewcraik** [comment]
I'm fine with this - @vijaysun-omr I think this should be good to test now. The scope has been reduced and the areas of concern delayed to another pull request.

**Comment by vijaysun-omr** [comment]
@genie-omr build all

**Comment by vijaysun-omr** [comment]
Since there are no more (partial) optimizer changes in this PR and the remaining changes just add helpers and an option, I am fine with it.

**Comment by vijaysun-omr** [comment]
@0xdaryl are you okay with the changes made ?

**Pull request merged**